### PR TITLE
RFC/WIP: FAPI

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
This is a preview / rfc / wip of the upcoming fapi API (new TCG spec).
It needs some more work, but I wanted to give people a way to play around with it.
See also the PR on tpm2-tools that provides CLI for playing around.
See also the PR on tpm2-pkcs11 that shows a use case.

Things I'd like to get some feedback on:
- How does the API feel ?
- How to add file locks for keystore and eventdir? Whole directory, per file, etc ?
- Does the logic for creating directories make sense ?
  We are using localstatedir and runstatedir and runstatedir is created via tpmfiles.d entrie.
  How does this work on non-systemd-tpmfiles.d systems ?

Other issues we're working on:
- More and faster testing
- Comments and documentation

I'll be force-pushing to this PR and trying to update the PR message as good as possible
until we can merge this into master and work in there. I'd give it 1-2 more weeks.

### CODE REVIEW

File: src/tss2-fapi/ifapi_profiles.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * All rights reserved.
+ ******************************************************************************/
+
+#ifndef IFAPI_PROFILES_H
+#define IFAPI_PROFILES_H
+
+#include "ifapi_io.h"
+#include "ifapi_policy_types.h"
+
+/** TODO
```
Comment (tstruk) [inline src/tss2-fapi/ifapi_profiles.h]: TODO what?

File: src/tss2-fapi/api/Fapi_AuthorizePolicy.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    size_t digest_idx;
+    IFAPI_Fapi_AuthorizePolicy *policy =
+        &context->cmd.Policy_AuthorizeNewPolicy;
+    TPMS_POLICYAUTHORIZATION *authorization = &policy->authorization;
+    TPMT_SIGNATURE *signature;
+    TPM2B_DIGEST aHash;
+    char *publicKey = NULL;
+
+    switch (context->state) {
+    case AUTHORIZE_NEW_LOAD_KEY:
```
Comment (tstruk) [inline src/tss2-fapi/api/Fapi_AuthorizePolicy.c]: In some cases we use the statecase macro and in some cases, like here, just case. Any reason for the inconsistency?

File: src/tss2-fapi/fapi_util.c [hunk 3]
```diff
@@ -0,0 +1,3782 @@
+/* SPDX-License-Identifier: BSD-2 */
```
Comment (tstruk) [inline src/tss2-fapi/fapi_util.c]: shouldn't this file be called ifapi_util.c?

File: src/tss2-fapi/fapi_util.c [hunk 4]
```diff
@@ -0,0 +1,3782 @@
...
+    case NV2_WRITE_WRITE_PREPARE:
+        /* Perform esys serialization if necessary */
+        r = ifapi_esys_serialize_object(context->esys, &context->nv_cmd.nv_object);
+        goto_if_error(r, "Prepare serialization", error_cleanup);
+
+        /* Start writing the NV object to the key store */
+        r = ifapi_keystore_store_async(&context->keystore, &context->io,
+                                       context->nv_cmd.nvPath,
+                                       &context->nv_cmd.nv_object);
+        goto_if
```
Comment (tstruk) [inline src/tss2-fapi/fapi_util.c]: Fix indentation

Comment (tstruk) [inline src/tss2-fapi/fapi_util.c]: This below only frees one object.

**Comment by tstruk** [comment]
@AndreasFuchsSIT could you resolve the conflicts please

**Comment by AndreasFuchsSIT** [comment]
Welcome back.
I'll be rebasing towards end of this / beginning of next week with some more code.

File: src/tss2-fapi/api/Fapi_AuthorizePolicy.c (hunk 2 above)
Comment (AndreasFuchsSIT) [reply src/tss2-fapi/api/Fapi_AuthorizePolicy.c]: statecase should now be everywhere

File: src/tss2-esys/esys_crypto_ossl.c [hunk 5]
```diff
@@ -607,6 +607,7 @@ iesys_cryptossl_pk_encrypt(TPM2B_PUBLIC * pub_tpm_key,
     if (1 != RSA_generate_key_ex(rsa_key,
                                  pub_tpm_key->publicArea.parameters.rsaDetail.keyBits,
                                  bne, NULL)) {
+        RAND_set_rand_method(rand_save);
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply src/tss2-esys/esys_crypto_ossl.c]: Correct

File: src/tss2-tcti/tcti-mssim.c [hunk 6]
```diff
@@ -319,9 +322,16 @@ tcti_mssim_receive (
     }
 
     if (timeout != TSS2_TCTI_TIMEOUT_BLOCK) {
-        LOG_WARNING ("Asynchronous I/O not implemented. The 'timeout' "
-                     "parameter must be TSS2_TCTI_TIMEOUT_BLOCK.");
-        return TSS2_TCTI_RC_BAD_VALUE;
+        LOG_TRACE ("Asynchronous I/O not actually implemented."
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply src/tss2-tcti/tcti-mssim.c]: put it into the ifdef TEST_FAPI_ASYNC that it is used for

File: src/tss2-fapi/fapi_util.c (hunk 4 above)
Comment (AndreasFuchsSIT) [reply src/tss2-fapi/fapi_util.c]: should be correct ?

**Comment by AndreasFuchsSIT** [comment]
Since the tools for FAPI now also compile, I'd like to merge this PR now, so the tools can also merge:
https://github.com/tpm2-software/tpm2-tools/pull/1778
Any further stuff will be separate PRs.

@tstruk Ok with you ?

File: Makefile.am [hunk 7]
```diff
@@ -382,6 +382,151 @@ src_tss2_rc_libtss2_rc_la_LDFLAGS = -Wl,--version-script=$(srcdir)/lib/tss2-rc.m
...
+    test/data/fapi/policy/pol_pcr16_0_or.json \
+    test/data/fapi/policy/pol_nv.json \
+    test/data/fapi/policy/pol_nv_written.json \
+    test/data/fapi/policy/pol_signed.json \
+    test/data/fapi/policy/pol_authorize.json \
+	test/data/fapi/policy/pol_authorize_nv.json \
+    test/data/fapi/policy/pol_secret.json \
+    test/data/fapi/policy/pol_password.json \
+    test/data/fapi/policy/pol_auth_value.json \
+	test/data/fapi/policy/pol_command_code.json \
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply Makefile.am]: Willdo. Damn \t

File: include/tss2/tss2_tpm2_types.h [hunk 8]
```diff
@@ -538,7 +540,7 @@ typedef UINT32 TPM2_PT;
 
 /* Definition of UINT32 TPM2_PT_PCR Constants <INOUT S> */
 typedef UINT32 TPM2_PT_PCR;
-#define TPM2_PT_TPM2_PCR_FIRST   ((TPM2_PT_PCR) 0x00000000) /* bottom of the range of TPM2_PT_PCR properties */
+#define TPM2_PT_PCR_FIRST        ((TPM2_PT_PCR) 0x00000000) /* bottom of the range of TPM2_PT_PCR properties */
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply include/tss2/tss2_tpm2_types.h]: Hmm... good point....
Shall I move this to backlock for 3.0 release ?

[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply include/tss2/tss2_tpm2_types.h]: Done

**Pull request merged**